### PR TITLE
Reduce hybrid inlines and reduce header dependencies

### PR DESCRIPTION
**Pull request by albinahlback**
With hybrid inlines I mean functions that are declared `static inline` except if called from `inlines.c`, such as
```c
FMPZ_INLINE
int fmpz_is_zero(const fmpz_t f)
{
    return (*f == 0);
}
```
Highlights:
- Make *very* internal hybrid inline functions to inline only
- Move fundamental `fmpz` macros and constants to `flint.h`
- Reduce header dependencies in `mpoly` headers
- Move type definition of `thread_pool_handle` to `flint.h` and remove all inclusions of `thread_pool.h` in header files such that `pthread.h` is never included in header files.

In total I get a 5% decrease in build time.

### CODE REVIEW

File: src/arb.h [hunk 1]
```diff
@@ -638,7 +638,7 @@ arb_sqr(arb_t res, const arb_t val, slong prec)
 
 /* vector functions */
 
-ARB_INLINE arb_ptr
+FLINT_FORCE_INLINE arb_ptr
 _arb_vec_entry_ptr(arb_ptr vec, slong i)
```
Comment (fredrik-johansson) [inline src/arb.h:642]: This is an example of a function that people call externally and which should thus remain in the library.

File: src/fmpq_mpoly.h [hunk 2]
```diff
@@ -202,9 +202,8 @@ int fmpq_mpoly_equal(const fmpq_mpoly_t A, const fmpq_mpoly_t B,
         && fmpz_mpoly_equal(A->zpoly, B->zpoly, ctx->zctx);
 }
 
-FMPQ_MPOLY_INLINE
-void fmpq_mpoly_swap(fmpq_mpoly_t A,
```
Comment (fredrik-johansson) [inline src/fmpq_mpoly.h:206]: I think we should keep the ``ctx`` object here for API uniformity.

File: src/arb_fmpz_poly.h [hunk 3]
```diff
@@ -46,7 +40,7 @@ ulong arb_fmpz_poly_deflation(const fmpz_poly_t input);
 
 void arb_fmpz_poly_complex_roots(acb_ptr roots, const fmpz_poly_t poly, int flags, slong target_prec);
 
-ARB_FMPZ_POLY_INLINE
+FLINT_FORCE_INLINE
 void arb_fmpz_poly_cos_minpoly(fmpz_poly_t res, ulong n)
```
Comment (fredrik-johansson) [inline src/arb_fmpz_poly.h:44]: Feel free to just remove this function instead.

File: src/acb_dirichlet.h [hunk 4]
```diff
@@ -190,13 +184,13 @@ void acb_dirichlet_backlund_s_bound(mag_t res, const arb_t t);
...
-ACB_DIRICHLET_INLINE void
+FLINT_FORCE_INLINE void
 acb_dirichlet_hardy_z_zero(arb_t res, const fmpz_t n, slong prec)
 {
     acb_dirichlet_hardy_z_zeros(res, n, 1, prec);
 }
 
-ACB_DIRICHLET_INLINE void
+FLINT_FORCE_INLINE void
 acb_dirichlet_zeta_zero(acb_t res, const fmpz_t n, slong prec)
```
Comment (fredrik-johansson) [inline src/acb_dirichlet.h:194]: This function is useful to export.

File: src/acb_dirichlet.h [hunk 5]
```diff
@@ -190,13 +184,13 @@ void acb_dirichlet_backlund_s_bound(mag_t res, const arb_t t);
 void acb_dirichlet_zeta_nzeros_gram(fmpz_t res, const fmpz_t n);
 slong acb_dirichlet_backlund_s_gram(const fmpz_t n);
 
-ACB_DIRICHLET_INLINE void
```
Comment (fredrik-johansson) [inline src/acb_dirichlet.h:193]: This function is useful to export.

**Comment by fredrik-johansson** [comment]
I'm a bit worried about not exporting ``swap`` methods which may well have outside use. For example [here is a use of ``arb_swap``](https://github.com/JeffreySarnoff/ArbNumerics.jl/blob/cea850c44a997993accf3c0c713d9d74a1ec47a5/src/libarb/ArbReal.jl#L71) in a non-Nemo Julia library.

But I guess we will see if people complain.

In general (apart from the comments) this looks OK.

File: src/acb_dirichlet.h (hunk 4 above)
Comment (albinahlback) [reply src/acb_dirichlet.h:194]: acb_dirichlet doesn't have `inlines.c`

**Comment by thofma** [comment]
If I understand correctly, this will remove a large chunk of functions from the library? Hm, how do you check if functions are called externally? Or is this all for flint 4.0?

**Comment by albinahlback** [comment]
I will be more conservative with the inline functions. I will also check against the larger libraries such as Macaulay2, MSOLVE and Singular.

**Comment by albinahlback** [comment]
> If I understand correctly, this will remove a large chunk of functions from the library? Hm, how do you check if functions are called externally? Or is this all for flint 4.0?
> 
> Edit: I personally don't think this is worth the 5% reduced build time.

The 5% reduction in build time shouldn't come from the inline functions, but rather the reduced header dependencies.

File: src/arb_fmpz_poly.h (hunk 3 above)
Comment (albinahlback) [reply src/arb_fmpz_poly.h]: Done!

**Comment by albinahlback** [comment]
I changed it to export swap functions and included context into mpoly swap functions. All symbols (apart from very internal functions and trivial `mpoly.h` functions) should remain.

**Comment by fredrik-johansson** [comment]
OK, this looks reasonable now.

**Comment by thofma** [comment]
Thanks Albin!

**Pull request merged**